streams/src/main/java/org/apache/kafka/streams/processor/internals/CompositeRestoreListener.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.processor.internals;
+
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.BatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.StateRestoreCallback;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+
+import java.util.Collection;
+
+class CompositeRestoreListener implements BatchingStateRestoreCallback, StateRestoreListener {
+
+    private final InternalStateRestoreAdapter stateRestoreAdapter;
+    private final StateRestoreListener storeRestoreListener;
+    private StateRestoreListener reportingStoreListener = new NoOpStateRestoreListener();
```

Comment:
Re synchronization: enforcing users to do sync themselves inside the function is fine, I did not see `implementors will need to provide synchronization if needed` so there may be a mis-understanding. Could you make that statement more clear in javadocs?

Re separating classes: One motivation I had is that, currently we did one "instanceof" for each call, and if the global listener is not set we still call a `no-op` listener; this does not seem optimized for me. Instead we can do a `null` check or even a boolean flag indicating if a global listener is ever set already. That would be more performant? If you can do that inside this composite class I think we could keep it centralized.